## Data Description
> In this phase, we describe the meanings of each column

The data descriptions can be found here: ("https://archive.ics.uci.edu/dataset/296/diabetes+130-us+hospitals+for+years+1999-2008")

In [1]:
# Import dependencies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import stats

In [2]:
# load data
df = pd.read_csv('/home/teofilo_acholla_ligawa_gafna/Documents/python_practice/collaborations/Diabetes_Readmissions/data/diabetic_data.csv')

# preview
df.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                 1  ...          No      No                   No   
1                 3  ...          No      Up                   No   
2                 2  ...          No      No                   No   
3                 2  ...          No      Up                   No   
4                 1  ...          No  Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                   No                        No                       No   
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   

   metformin-pioglitazone  change diabetesMed readmitted  
0                      No      No          No         NO  
1                      No      Ch         Yes        >30  
2                      No      No         Yes         NO  
3                      No      Ch         Yes         NO  
4                      No      Ch         Yes         NO  

[5 rows x 50 columns]

In [3]:
# preview tail for consistency
df.tail()

encounter_id  patient_nbr             race  gender      age weight  \
101761     443847548    100162476  AfricanAmerican    Male  [70-80)      ?   
101762     443847782     74694222  AfricanAmerican  Female  [80-90)      ?   
101763     443854148     41088789        Caucasian    Male  [70-80)      ?   
101764     443857166     31693671        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
101761                  1                         3                    7   
101762                  1                         4                    5   
101763                  1                         1                    7   
101764                  2                         3                    7   
101765                  1                         1                    7   

        time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
101761                 3  ...          No    Down                   No   
101762                 5  ...          No  Steady                   No   
101763                 1  ...          No    Down                   No   
101764                10  ...          No      Up                   No   
101765                 6  ...          No      No                   No   

        glipizide-metformin  glimepiride-pioglitazone  \
101761                   No                        No   
101762                   No                        No   
101763                   No                        No   
101764                   No                        No   
101765                   No                        No   

        metformin-rosiglitazone  metformin-pioglitazone  change diabetesMed  \
101761                       No                      No      Ch         Yes   
101762                       No                      No      No         Yes   
101763                       No                      No      Ch         Yes   
101764                       No                      No      Ch         Yes   
101765                       No                      No      No          No   

       readmitted  
101761        >30  
101762         NO  
101763         NO  
101764         NO  
101765         NO  

[5 rows x 50 columns]

## Preliminary Data Inspection
> In the phase, the study shall seek to undersatand the 'data' about the data.

In [4]:
# metadata
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

In [21]:
def inspect_df(df):
    """This prints out the summary from the inspection of the data"""

    return {"Dimensions": f"This data set has {df.shape[0]} rows and {df.shape[1]} columns",
            "Duplicates": f"The data has {df.duplicated().sum()} duplicated entries and {len(df) - df.duplicated().sum()} non duplicated entries",
            "Missing values (%)": f"{(df.isna().sum().sum() / (df.shape[0] * df.shape[1])) * 100} % of the data has missing values",
            "Summary statistics": df.describe().T,
            "Info (printed above)":df.info()}

def inspect_col(df, col):
  """Summary Inspection of the column in focus"""
  
  return {"Top 5 value counts": df[col].value_counts()[:5],
          "Number of unique values": df[col].nunique(),
          "data type": df[col].dtype,
          "Missing Values": f"{col} has {df[col].isna().sum()} missing values. This equates to {round(df[col].isna().sum()/df.__len__(),5) * 100} % missing"}
  
  
  # Summary data of the data frame
inspect_df(df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      99493 non-null   object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    3197 non-null    object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                61510 non-null   object
 11  medical_specialty         51817 non-null   object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

{'Dimensions': 'This data set has 101766 rows and 50 columns',
 'Duplicates': 'The data has 0 duplicated entries and 101766 non duplicated entries',
 'Missing values (%)': '3.790047756618124 % of the data has missing values',
 'Summary statistics':                              count          mean           std      min  \
 encounter_id              101766.0  1.652016e+08  1.026403e+08  12522.0   
 patient_nbr               101766.0  5.433040e+07  3.869636e+07    135.0   
 admission_type_id         101766.0  2.024006e+00  1.445403e+00      1.0   
 discharge_disposition_id  101766.0  3.715642e+00  5.280166e+00      1.0   
 admission_source_id       101766.0  5.754437e+00  4.064081e+00      1.0   
 time_in_hospital          101766.0  4.395987e+00  2.985108e+00      1.0   
 num_lab_procedures        101766.0  4.309564e+01  1.967436e+01      1.0   
 num_procedures            101766.0  1.339730e+00  1.705807e+00      0.0   
 num_medications           101766.0  1.602184e+01  8.127566e+00     

In [17]:
# making summary statistics clear
df.describe()

encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    4.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count       101766.000000   101766.000000    101766.000000      101766.000000   
mean            43.095641        1.339730        16.021844           0.369357   
std             19.674362        1.705807         8.127566           1.267265   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        10.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses  
count     101766.000000     101766.000000     101766.000000  
mean           0.197836          0.635566          7.422607  
std            0.930472          1.262863          1.933600  
min            0.000000          0.000000          1.000000  
25%            0.000000          0.000000          6.000000  
50%            0.000000          0.000000          8.000000  
75%            0.000000          1.000000          9.000000  
max           76.000000         21.000000         16.000000

The findings from above indicate that there is no missing value yet upon inspection of the first five rows, there were question marks which do not provide us with any meaningful information thus the study considers them as missing values and to regard them as such, they shall be replaced by NaN.

In [18]:
# replacing with NAN
df.replace('?', np.nan, inplace=True)

# Preview
df.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)    NaN   
1        149190     55629189        Caucasian  Female  [10-20)    NaN   
2         64410     86047875  AfricanAmerican  Female  [20-30)    NaN   
3        500364     82442376        Caucasian    Male  [30-40)    NaN   
4         16680     42519267        Caucasian    Male  [40-50)    NaN   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                 1  ...          No      No                   No   
1                 3  ...          No      Up                   No   
2                 2  ...          No      No                   No   
3                 2  ...          No      Up                   No   
4                 1  ...          No  Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                   No                        No                       No   
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   

   metformin-pioglitazone  change diabetesMed readmitted  
0                      No      No          No         NO  
1                      No      Ch         Yes        >30  
2                      No      No         Yes         NO  
3                      No      Ch         Yes         NO  
4                      No      Ch         Yes         NO  

[5 rows x 50 columns]

In [19]:
# Inspect the data once more
inspect_df(df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      99493 non-null   object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    3197 non-null    object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                61510 non-null   object
 11  medical_specialty         51817 non-null   object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

{'Dimensions': 'This data set has 101766 rows and 50 columns',
 'Duplicates': 'The data has 0 duplicated entries and 101766 non duplicated entries',
 'Missing values (%)': '3.790047756618124 % of the data has missing values',
 'Summary statistics':                              count          mean           std      min  \
 encounter_id              101766.0  1.652016e+08  1.026403e+08  12522.0   
 patient_nbr               101766.0  5.433040e+07  3.869636e+07    135.0   
 admission_type_id         101766.0  2.024006e+00  1.445403e+00      1.0   
 discharge_disposition_id  101766.0  3.715642e+00  5.280166e+00      1.0   
 admission_source_id       101766.0  5.754437e+00  4.064081e+00      1.0   
 time_in_hospital          101766.0  4.395987e+00  2.985108e+00      1.0   
 num_lab_procedures        101766.0  4.309564e+01  1.967436e+01      1.0   
 num_procedures            101766.0  1.339730e+00  1.705807e+00      0.0   
 num_medications           101766.0  1.602184e+01  8.127566e+00     

In [22]:
# Inspecting each column
for col in df.columns:
    print(col)
    print(inspect_col(df, col))
    print("="*70)

encounter_id
{'Top 5 value counts': 2278392      1
190792044    1
190790070    1
190789722    1
190786806    1
Name: encounter_id, dtype: int64, 'Number of unique values': 101766, 'data type': dtype('int64'), 'Missing Values': 'encounter_id has 0 missing values. This equates to 0.0 % missing'}
patient_nbr
{'Top 5 value counts': 88785891    40
43140906    28
1660293     23
88227540    23
23199021    23
Name: patient_nbr, dtype: int64, 'Number of unique values': 71518, 'data type': dtype('int64'), 'Missing Values': 'patient_nbr has 0 missing values. This equates to 0.0 % missing'}
race
{'Top 5 value counts': Caucasian          76099
AfricanAmerican    19210
Hispanic            2037
Other               1506
Asian                641
Name: race, dtype: int64, 'Number of unique values': 5, 'data type': dtype('O'), 'Missing Values': 'race has 2273 missing values. This equates to 2.234 % missing'}
gender
{'Top 5 value counts': Female             54708
Male               47055
Unknown/Invalid  

### Observations:
* There are 101,766 unique records with 71518 unique patients - This shows that it is possible that at least one or more patients has had at least 2 or more encounters that has forced them to be hospitalized.
* Missing values form 3.79 % of this data and there are no duplicates in the data.
* There are columns that have been given wrong types. For instance, columns like `encounter_id` and `patient_nbr` are taken as integers yet they are primary identifiers of the encounter and the patient in question respectively.

* `gender` has a category known as 'Unknown/Invalid' which hints at missing values.

* `weight` has over 96% of the column missing.
* `admission_type_id`, `discharge_disposition_id` and `admission_source_id` are not really IDs in the conventional sense but are more like categorical variables which are nominal in nature- these features will be type casted to category.

* `payer_code` has about 40% missing values and `medical_specialty` has about 50% missing values.

* Other columns such as `race`, `diag_1`, `diag_2` and `diag_3` each have a small number of missing values with none of them having more than 3 % missing - these can be easily dropped. It is worth pointing out that these columns are supposed to be categorical columns because they provide information on the diagnoses given to the patient while taking on a limited number of categories without any inherent order. The study referred to these files for better understanding: 'https://github.com/WendyMwiti/Diabetes_Readmissions/blob/7a78a50202d6044b16bc601093845088f86266be/data/references'

* Columns 24 to 46 are drug names. They are to be casted as category as they contain few categorical values that are repetitive.

* Features of type object that are supposed to be object will be type casted to category.

* The target column in this task is `readmitted` which tells us whether a patient was readmitted within 30 days, greater than 30 days or not readmitted at all.

* The rows with consideration of the column `discharge_disposition_id` that hint towards the fact that the patient was taken to hospice or death will be removed as they do not help the model - they cannot be readmitted. They are denoted by values 11,13,14,19,20 and 21.
